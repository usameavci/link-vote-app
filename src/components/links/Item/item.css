.link-item {
  display: grid;
  padding: 0.5rem;
  grid-gap: 1rem;
  position: relative;
  border-radius: 4px;
  align-items: center;
  justify-content: flex-start;
  border: 1px solid transparent;
  grid-template-columns: auto 1fr;
}

.link-item-vote {
  width: 80px;
  height: 80px;
  display: flex;
  border-radius: 4px;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  background-color: #ececec;
  border: 1px solid #c1c1c1;
}

.link-item-vote .vote-point {
  font-weight: 700;
  font-size: 1.75rem;
}

.link-item-vote .vote-label {
  font-size: 0.75rem;
}

.link-item-title {
  font-weight: 700;
  font-size: 1.25rem;
}

.link-item-url {
  color: #7f7f7f;
  margin-top: 0.25rem;
}

.link-item-actions {
  display: flex;
  margin-top: 1rem;
  align-items: flex-end;
  justify-content: space-between;
}

.link-item-actions .link-item-action {
  border: 0;
  border-radius: 3px;
  padding: 0.25rem 0.5rem;
  background-color: #f5f5f5;
}

.link-item-actions .link-item-action:hover {
  background-color: #eee;
}

.link-item .link-item-remove {
  border: 0;
  color: #fff;
  top: -0.5rem;
  display: flex;
  width: 1.5rem;
  height: 1.5rem;
  right: -0.5rem;
  visibility: hidden;
  border-radius: 50%;
  position: absolute;
  align-items: center;
  justify-content: center;
  background-color: #a10007;
}

.link-item:hover {
  border-color: #eaeaea;
  background-color: #f7f7f7;
}

.link-item:hover .link-item-remove {
  visibility: visible;
}
