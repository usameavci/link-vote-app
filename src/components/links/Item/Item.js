import upperCase from "lodash/upperCase";
import React, { useContext } from "react";

import Notification from "../../../utils/Notification";
import { LinkContext } from "../../../contexts/LinkContext";

import "./item.css";

const Item = (props) => {
  const { deleteLink, voteLink } = useContext(LinkContext);

  const handleDeleteLink = (e) => {
    Notification.prompt(
      "Remove Link",
      `Do you want to remove:<br><b>${upperCase(props.link.title)}</b>`,
      (notification) => {
        deleteLink(props.link).then(() => {
          notification.close();

          Notification.success(`${upperCase(props.link.title)} removed`);
        });
      }
    );

    return e.preventDefault();
  };

  return (
    <div data-testid="link-item" className="link-item">
      <button data-testid="link-item-remove" className="link-item-remove" onClick={handleDeleteLink}>
        <i className="fa fa-times" />
      </button>
      <div className="link-item-vote">
        <div className="vote-point">{props.link.vote}</div>
        <div className="vote-label">POINTS</div>
      </div>
      <div className="link-item-body">
        <div className="link-item-title">{props.link.title}</div>
        <div className="link-item-url">({props.link.url})</div>
        <div className="link-item-actions">
          <button
            data-testid="link-item-up-vote"
            className="link-item-action"
            onClick={() => voteLink(props.link, "up")}
          >
            <i className="fa fa-arrow-up" /> Up Vote
          </button>
          <button
            data-testid="link-item-down-vote"
            className="link-item-action"
            onClick={() => voteLink(props.link, "down")}
          >
            <i className="fa fa-arrow-down" /> Down Vote
          </button>
        </div>
      </div>
    </div>
  );
};

export default Item;
