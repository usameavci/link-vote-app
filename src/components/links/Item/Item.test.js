import React from "react";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

import Item from "./Item";
import LinkContextProvider from "../../../contexts/LinkContext";

describe("Index page component tests", () => {
  const link = {
    title: "Google",
    url: "https://google.com",
    vote: 0,
    createdAt: "2020-04-12T15:38:53.412Z",
    updatedAt: "2020-04-12T15:38:53.412Z",
    id: "dbde0208-9f9e-2304-2cfe-d5af1359b1b4",
  };

  it("renders without crashing", async () => {
    const component = (
      <Router>
        <LinkContextProvider>
          <Item link={link} />
        </LinkContextProvider>
      </Router>
    );

    let getByTestId;

    await act(async () => ({ getByTestId } = await render(component)));

    expect(getByTestId("link-item")).toBeInTheDocument();
  });

  it("has up vote, down vote and remove buttons", async () => {
    const component = (
      <Router>
        <LinkContextProvider>
          <Item link={link} />
        </LinkContextProvider>
      </Router>
    );

    let getByTestId;

    await act(async () => ({ getByTestId } = await render(component)));

    expect(getByTestId("link-item-remove")).toBeInTheDocument();
    expect(getByTestId("link-item-up-vote")).toBeInTheDocument();
    expect(getByTestId("link-item-down-vote")).toBeInTheDocument();
  });

  it("matches snapshots", async () => {
    const component = (
      <Router>
        <LinkContextProvider>
          <Item link={link} />
        </LinkContextProvider>
      </Router>
    );

    let tree;

    await renderer.act(async () => (tree = renderer.create(component)));

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
