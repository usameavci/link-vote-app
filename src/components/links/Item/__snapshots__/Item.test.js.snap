// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page component tests matches snapshots 1`] = `
<div
  className="link-item"
  data-testid="link-item"
>
  <button
    className="link-item-remove"
    data-testid="link-item-remove"
    onClick={[Function]}
  >
    <i
      className="fa fa-times"
    />
  </button>
  <div
    className="link-item-vote"
  >
    <div
      className="vote-point"
    >
      0
    </div>
    <div
      className="vote-label"
    >
      POINTS
    </div>
  </div>
  <div
    className="link-item-body"
  >
    <div
      className="link-item-title"
    >
      Google
    </div>
    <div
      className="link-item-url"
    >
      (
      https://google.com
      )
    </div>
    <div
      className="link-item-actions"
    >
      <button
        className="link-item-action"
        data-testid="link-item-up-vote"
        onClick={[Function]}
      >
        <i
          className="fa fa-arrow-up"
        />
         Up Vote
      </button>
      <button
        className="link-item-action"
        data-testid="link-item-down-vote"
        onClick={[Function]}
      >
        <i
          className="fa fa-arrow-down"
        />
         Down Vote
      </button>
    </div>
  </div>
</div>
`;
